src/linear_algebra/matrix.lean

```diff
@@ -200,24 +202,32 @@ def diag : (matrix n n M) →ₗ[R] n → M := {
   smul   := by { intros, ext, refl, } }
 
 @[simp] lemma diag_one [decidable_eq n] :
-  (diag : (matrix n n R) →ₗ[R] n → R) 1 = λ i, 1 := by {
-    dunfold diag, ext, simp [one_val_eq], }
+  diag n R R 1 = λ i, 1 := by { dunfold diag, ext, simp [one_val_eq],}
 
-protected lemma smul_sum {α : Type u} {s : finset α} {f : α → M} {c : R} :
-  c • s.sum f = s.sum (λx, c • f x) := (s.sum_hom _).symm
+@[simp] lemma diag_transpose (A : matrix n n R) : diag n R R Aᵀ = diag n R R A := rfl
 
 /--
 The trace of a square matrix.
 -/
 def trace : (matrix n n M) →ₗ[R] M := {
-  to_fun := finset.univ.sum ∘ (diag : (matrix n n M) →ₗ[R] n → M),
+  to_fun := finset.univ.sum ∘ (diag n R M),
   add    := by { intros, apply finset.sum_add_distrib, },
-  smul   := by { intros, simp [matrix.smul_sum], } }
+  smul   := by { intros, simp [finset.smul_sum], } }
 
 @[simp] lemma trace_one [decidable_eq n] :
-  (trace : (matrix n n R) →ₗ[R] R) 1 = fintype.card n := by {
-    have h : @trace n _ R R _ _ _ 1 = finset.univ.sum (@diag n _ R R _ _ _ 1) := rfl,
-    rw [h, diag_one, finset.sum_const, add_monoid.smul_one], refl, }
+  trace n R R 1 = fintype.card n := by {
```

Comment:
The `by { ... }` pattern should be reserverd for one-liners.
Here, I think a natural way to save it is to put the `have` in term-mode, before the `by`, and then prove the result with `by rw [...]` (without `{}`)

Reply:
Thanks, I have moved the `have` outside of the tactics area but I have had to keep the `{}` to allow two tactics after the `by` since I still need `refl` after the `rw`.

Reply:
You can skip the `{}` if you separate the tactics with `;` instead of `,`. (This has different behavior for tactics that produce multiple goals, but for straight line tactics this is another way to write these sequences.)

Reply:
Thanks @digama0 

Actually I used to be a heavy user (and abuser) of `;`. Some time ago @jcommelin pointed out to me that injudicious use of `;` can make it difficult to read / step through proofs, so I trained myself away from using it but perhaps here it would be better?